server/src/main/java/io/unitycatalog/server/handler/CoordinatedCommitsHandler.java

```diff
@@ -0,0 +1,140 @@
+
+  public static void validate(Commit commit) {
+    // Validate the commit object
+    assert commit.getTableId() != null;
+    //    assert commit.getTableUri() != null;
+
+    // TODO: Add other assertions like the table URI path exists
+
+    // Validate the commit info object
+    if (commit.getCommitInfo() != null) {
+      assert commit.getCommitInfo().getIsDisownCommit() != null;
+      assert commit.getCommitInfo().getFileSize() > 0;
+      assert commit.getCommitInfo().getFileName() != null;
+      assert commit.getCommitInfo().getVersion() != null;
+      assert commit.getCommitInfo().getTimestamp() != null;
+      assert commit.getCommitInfo().getFileModificationTimestamp() != null;
+    } else {
+      // If commit info is null, then it should be a backfill only commit
+      assert commit.getLatestBackfilledVersion() != null;
+    }
+  }
+
+  public static void validateTablePath(Commit commit) {
+    TableInfo tableInfo = TABLE_REPOSITORY.getTableById(commit.getTableId());
+    //    assert tableInfo.getTableType() == TableType.MANAGED;
+    assert tableInfo.getDataSourceFormat() == DataSourceFormat.DELTA;
+
+    // TODO: Add other assertions like verifying the table path (tableInfo.getStorageLocation)
+  }
+
+  public static void validateOnboardingCommit(Commit commit) {
+    assert commit.getCommitInfo() != null;
+    // Onboarding commit cannot be a disown commit
+    assert !commit.getCommitInfo().getIsDisownCommit();
+  }
+
+  public static void handleFirstCommit(Commit commit) {
+    validateOnboardingCommit(commit);
+    COMMIT_REPOSITORY.saveCommit(commit);
+  }
+
+  public static void handleBackfillOnlyCommit(
+      String tableId, Long latestBackfilledVersion, CommitDAO firstCommit, CommitDAO lastCommit) {
+    // Nothing to delete, return
+    if (latestBackfilledVersion < firstCommit.getCommitVersion()) {
+      return;
+    }
+    // We only need to delete when there is more than one commit. We always keep the last commit.
+    if (firstCommit.getCommitVersion() < lastCommit.getCommitVersion()) {
+      COMMIT_REPOSITORY.backfillCommits(
+          UUID.fromString(tableId),
+          Math.min(latestBackfilledVersion, lastCommit.getCommitVersion() - 1),
+          firstCommit,
+          lastCommit.getCommitVersion());
+    }
+    if (latestBackfilledVersion == lastCommit.getCommitVersion()) {
+      // Mark the last commit as the latest backfilled version
+      COMMIT_REPOSITORY.markCommitAsLatestBackfilled(
+          UUID.fromString(tableId), lastCommit.getCommitVersion());
+    }
+
+    // TODO: Should we also retain the disown commit if it's backfilled?
+  }
+
+  public static void handleReboardCommit() {
+    throw new RuntimeException("Reboarding of tables not allowed!");
+  }
+
+  public static void handleNormalCommit(
+      Commit commit, CommitDAO firstCommit, CommitDAO lastCommit) {
+    if (commit.getCommitInfo().getVersion() <= lastCommit.getCommitVersion()) {
```

Comment:
Shouldn't all these checks and commit to table be in a single transaction?

Reply:
yeah that's true, fixed now